File: telugu_panchangam/personal/election_assessors/graha_nature.py
"""Graha-nature predicates, separate from chart geometry."""

from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass
from math import floor
from typing import Any, Literal

from ...panchangam_names import RASHI_NAMES
from ..election_chart_rules import ELECTION_CHART_PLANETS
from .contracts import COMPLETE_GRAHA_FACTS_UNAVAILABLE, PrimitiveOutcome
from .conventions import ELECTION_CHART_CONVENTIONS
from .event_admission import PlanetPosition, planet_positions

NATURAL_MALEFIC_LUNAR_PHASE_GUARD_DEGREES = 0.02
_BOUNDARY_EPSILON = 1e-9

NATURAL_GRAHA_NATURE_CONVENTION_ID = (
    'phaladeepika-natural-graha-nature-whole-sign-v1'
)
_NATURAL_GRAHA_NATURE_CONVENTION = ELECTION_CHART_CONVENTIONS[
    NATURAL_GRAHA_NATURE_CONVENTION_ID
]
_FIXED_MALEFICS = frozenset(
    _NATURAL_GRAHA_NATURE_CONVENTION['fixed_malefics']
)
_FIXED_BENEFICS = frozenset(
    _NATURAL_GRAHA_NATURE_CONVENTION['fixed_benefics']
)
_PHASE_QUANTIZATION_SCALE = 10 ** int(
    _NATURAL_GRAHA_NATURE_CONVENTION['phase_quantization_decimal_places']
)
_PHASE_GUARD_UNITS = floor(
    NATURAL_MALEFIC_LUNAR_PHASE_GUARD_DEGREES
    * _PHASE_QUANTIZATION_SCALE
    + 0.5
)
_HALF_CIRCLE_UNITS = 180 * _PHASE_QUANTIZATION_SCALE
_FULL_CIRCLE_UNITS = 360 * _PHASE_QUANTIZATION_SCALE

NaturalGrahaNature = Literal['benefic', 'malefic', 'unknown']


@dataclass(frozen=True)
class NaturalGrahaNatureResult:
    """Resolved natural natures for one complete canonical chart."""

    complete: bool
    natures: Mapping[str, NaturalGrahaNature]
    evidence: tuple[str, ...] = ()


def _longitude(position: PlanetPosition) -> float:
    return RASHI_NAMES.index(position.rashi) * 30 + position.degree


def _unknown_natures() -> dict[str, NaturalGrahaNature]:
    return {name: 'unknown' for name in ELECTION_CHART_PLANETS}


def _resolved_chandra_nature(
    positions: Mapping[str, PlanetPosition],
) -> tuple[NaturalGrahaNature, float]:
    raw_elongation = (
        _longitude(positions['Chandra']) - _longitude(positions['Surya'])
    ) % 360
    elongation_units = floor(
        raw_elongation * _PHASE_QUANTIZATION_SCALE + 0.5
    )
    elongation = elongation_units / _PHASE_QUANTIZATION_SCALE
    boundary_distance_units = min(
        elongation_units,
        abs(elongation_units - _HALF_CIRCLE_UNITS),
        abs(_FULL_CIRCLE_UNITS - elongation_units),
    )
    if boundary_distance_units <= _PHASE_GUARD_UNITS:
        return 'unknown', elongation
    if elongation_units < _HALF_CIRCLE_UNITS:
        return 'benefic', elongation
    return 'malefic', elongation


def classify_natural_graha_natures(
    chart: Mapping[str, Any],
) -> NaturalGrahaNatureResult:
    """Classify all nine grahas under the disclosed natural-nature v1."""
    positions = planet_positions(chart) if isinstance(chart, Mapping) else None
    if positions is None:
        return NaturalGrahaNatureResult(
            complete=False,
            natures=_unknown_natures(),
            evidence=(
                'Complete canonical nine-graha facts are unavailable or invalid.',
            ),
        )

    natures: dict[str, NaturalGrahaNature] = _unknown_natures()
    for name in _FIXED_MALEFICS:
        natures[name] = 'malefic'
    for name in _FIXED_BENEFICS:
        natures[name] = 'benefic'

    chandra_nature, elongation = _resolved_chandra_nature(positions)
    natures['Chandra'] = chandra_nature
    if chandra_nature == 'unknown':
        chandra_evidence = (
            f'Chandra elongation {elongation:.4f}° is inside the inclusive '
            f'{NATURAL_MALEFIC_LUNAR_PHASE_GUARD_DEGREES:.2f}° '
            'phase-boundary guard.'
        )
    else:
        chandra_evidence = (
            f'Chandra elongation {elongation:.4f}° resolves {chandra_nature}.'
        )

    budha = positions['Budha']
    companions = [
        name
        for name in ELECTION_CHART_PLANETS
        if name != 'Budha' and positions[name].rashi == budha.rashi
    ]
    malefic_witnesses = [
        name for name in companions if natures[name] == 'malefic'
    ]
    if malefic_witnesses:
        natures['Budha'] = 'malefic'
        budha_evidence = (
            f'Budha shares {budha.rashi} with resolved malefic '
            f'{", ".join(malefic_witnesses)}.'
        )
    elif 'Chandra' in companions and natures['Chandra'] == 'unknown':
        natures['Budha'] = 'unknown'
        budha_evidence = (
            f'Budha shares {budha.rashi} only with phase-unknown Chandra '
            'as a possible malefic witness.'
        )
    else:
        natures['Budha'] = 'benefic'
        budha_evidence = (
            f"No resolved malefic shares Budha's sidereal Rasi {budha.rashi}."
        )

    return NaturalGrahaNatureResult(
        complete=True,
        natures={name: natures[name] for name in ELECTION_CHART_PLANETS},
        evidence=(chandra_evidence, budha_evidence),
    )


def _house_set(houses: Any) -> frozenset[int] | None:
    if isinstance(houses, (str, bytes)):
        return None
    try:
        values = tuple(houses)
    except TypeError:
        return None
    if not values or any(
        type(house) is not int or not 1 <= house <= 12 for house in values
    ):
        return None
    return frozenset(values)


def _house_witnesses(
    chart: Mapping[str, Any], houses: Any
) -> tuple[NaturalGrahaNatureResult, list[tuple[str, int]]] | None:
    target_houses = _house_set(houses)
    if target_houses is None:
        return None
    result = classify_natural_graha_natures(chart)
    if not result.complete:
        return result, []
    positions = planet_positions(chart)
    if positions is None:
        return NaturalGrahaNatureResult(False, _unknown_natures()), []
    occupants = [
        (name, positions[name].house)
        for name in ELECTION_CHART_PLANETS
        if positions[name].house in target_houses
    ]
    return result, occupants


def evaluate_existential_benefic_house_set(
    chart: Mapping[str, Any], houses: Any
) -> PrimitiveOutcome:
    """Pass when any target house has a resolved natural benefic witness."""
    witnesses = _house_witnesses(chart, houses)
    if witnesses is None:
        return PrimitiveOutcome('unknown', ('The required house set is invalid.',))
    result, occupants = witnesses
    if not result.complete:
        return PrimitiveOutcome('unknown', result.evidence)
    benefics = [
        (name, house)
        for name, house in occupants
        if result.natures[name] == 'benefic'
    ]
    if benefics:
        detail = ', '.join(f'{name} (house {house})' for name, house in benefics)
        return PrimitiveOutcome(
            'pass', (f'Resolved natural-benefic witness: {detail}.',)
        )
    unresolved = [
        (name, house)
        for name, house in occupants
        if result.natures[name] == 'unknown'
    ]
    if unresolved:
        detail = ', '.join(f'{name} (house {house})' for name, house in unresolved)
        return PrimitiveOutcome(
            'unknown', (f'Natural-benefic witness remains unresolved: {detail}.',)
        )
    return PrimitiveOutcome(
        'fail', ('No resolved natural benefic occupies the required house set.',)
    )


def evaluate_forbidden_malefic_house_set(
    chart: Mapping[str, Any], houses: Any
) -> PrimitiveOutcome:
    """Pass only when no target house has a natural-malefic witness."""
    witnesses = _house_witnesses(chart, houses)
    if witnesses is None:
        return PrimitiveOutcome('unknown', ('The forbidden house set is invalid.',))
    result, occupants = witnesses
    if not result.complete:
        return PrimitiveOutcome('unknown', result.evidence)
    malefics = [
        (name, house)
        for name, house in occupants
        if result.natures[name] == 'malefic'
    ]
    if malefics:
        detail = ', '.join(f'{name} (house {house})' for name, house in malefics)
        return PrimitiveOutcome(
            'fail', (f'Forbidden natural-malefic witness: {detail}.',)
        )
    unresolved = [
        (name, house)
        for name, house in occupants
        if result.natures[name] == 'unknown'
    ]
    if unresolved:
        detail = ', '.join(f'{name} (house {house})' for name, house in unresolved)
        return PrimitiveOutcome(
            'unknown', (f'Forbidden-malefic check remains unresolved: {detail}.',)
        )
    return PrimitiveOutcome(
        'pass', ('No resolved natural malefic occupies the forbidden house set.',)
    )


def evaluate_house_free_of_natural_malefics(
    rule: Mapping[str, Any],
    positions: Mapping[str, PlanetPosition] | None,
    *,
    house_frame_uncertain: bool = False,
) -> PrimitiveOutcome:
    """Apply the disclosed Annaprasana natural-malefic convention."""
    if positions is None:
        return PrimitiveOutcome('unknown', (COMPLETE_GRAHA_FACTS_UNAVAILABLE,))
    if house_frame_uncertain:
        return PrimitiveOutcome(
            'unknown', ('The local-Lagna house frame is uncertain.',))

    house = rule.get('house')
    fixed = [
        name for name in rule.get('fixed_malefics', ())
        if positions[name].house == house
    ]
    if fixed:
        return PrimitiveOutcome(
            'fail',
            (f'Natural malefics in Lagna: {", ".join(fixed)}.',),
        )

    chandra = positions['Chandra']
    if chandra.house != house:
        return PrimitiveOutcome(
            'pass',
            ('Natural malefics in Lagna: none; Chandra is outside Lagna.',),
        )

    surya = positions['Surya']
    elongation = (_longitude(chandra) - _longitude(surya)) % 360
    guard = rule.get(
        'lunar_phase_guard_degrees',
        NATURAL_MALEFIC_LUNAR_PHASE_GUARD_DEGREES,
    )
    boundary_distance = min(
        elongation, 360 - elongation, abs(elongation - 180))
    if boundary_distance <= guard + _BOUNDARY_EPSILON:
        return PrimitiveOutcome(
            'unknown',
            (
                f'Chandra occupies Lagna at {elongation:.2f}\N{DEGREE SIGN} '
                f'solar elongation, inside the disclosed ±{guard:.2f}° '
                'phase boundary guard.',
            ),
        )
    if elongation > 180:
        return PrimitiveOutcome(
            'fail',
            (
                f'Natural malefics in Lagna: waning Chandra '
                f'({elongation:.2f}\N{DEGREE SIGN} solar elongation).',
            ),
        )
    return PrimitiveOutcome(
        'pass',
        (
            f'Natural malefics in Lagna: none; waxing Chandra '
            f'({elongation:.2f}\N{DEGREE SIGN} solar elongation) is not '
            'malefic under this convention.',
        ),
    )

File: src/scorer/election-assessors/graha-nature.ts
import { RASI_NAMES } from '../../data/rasis';
import rulesContract from '../../data/election-chart-rules.generated.json';
import type { ElectionChartSnapshot } from '../../lib/election-chart-api';
import {
  COMPLETE_GRAHA_FACTS_UNAVAILABLE,
  type ElectionPrimitiveRule,
  type PlanetPosition,
  type PrimitiveOutcome,
} from './contracts';
import { completePlanetPositions } from './event-admission';

const NATURAL_MALEFIC_LUNAR_PHASE_GUARD_DEGREES = 0.02;
const BOUNDARY_EPSILON = 1e-9;

export type NaturalGrahaNature = 'benefic' | 'malefic' | 'unknown';

export interface NaturalGrahaNatureResult {
  complete: boolean;
  natures: Readonly<Record<string, NaturalGrahaNature>>;
  evidence: string[];
}

interface NaturalGrahaNatureConvention {
  fixed_malefics: string[];
  fixed_benefics: string[];
  chandra_phase_guard_degrees: number;
  phase_quantization_decimal_places: number;
  phase_quantization_rounding: 'half_up_nonnegative';
  budha_association: 'same_sidereal_rashi';
  house_system: 'whole_sign';
}

export const NATURAL_GRAHA_NATURE_CONVENTION_ID =
  'phaladeepika-natural-graha-nature-whole-sign-v1' as const;
const NATURAL_GRAHA_NATURE_CONVENTION = (
  rulesContract.conventions as unknown as Record<string, NaturalGrahaNatureConvention>
)[NATURAL_GRAHA_NATURE_CONVENTION_ID];
const CANONICAL_PLANET_ORDER = [...rulesContract.vacancy_includes];
const CANONICAL_PLANETS = new Set(CANONICAL_PLANET_ORDER);
const FIXED_MALEFICS = new Set(NATURAL_GRAHA_NATURE_CONVENTION.fixed_malefics);
const FIXED_BENEFICS = new Set(NATURAL_GRAHA_NATURE_CONVENTION.fixed_benefics);
const PHASE_QUANTIZATION_SCALE = 10 ** (
  NATURAL_GRAHA_NATURE_CONVENTION.phase_quantization_decimal_places
);
const PHASE_GUARD_UNITS = Math.floor(
  NATURAL_GRAHA_NATURE_CONVENTION.chandra_phase_guard_degrees
  * PHASE_QUANTIZATION_SCALE + 0.5,
);
const HALF_CIRCLE_UNITS = 180 * PHASE_QUANTIZATION_SCALE;
const FULL_CIRCLE_UNITS = 360 * PHASE_QUANTIZATION_SCALE;

function longitude(position: PlanetPosition): number {
  return RASI_NAMES.indexOf(position.rashi) * 30 + position.degree;
}

function unknownNatures(): Record<string, NaturalGrahaNature> {
  return Object.fromEntries(
    CANONICAL_PLANET_ORDER.map(name => [name, 'unknown' as const]),
  );
}

function resolvedChandraNature(
  positions: ReadonlyMap<string, PlanetPosition>,
): [NaturalGrahaNature, number] {
  const raw = (
    (longitude(positions.get('Chandra')!) - longitude(positions.get('Surya')!))
    % 360 + 360
  ) % 360;
  const elongationUnits = Math.floor(raw * PHASE_QUANTIZATION_SCALE + 0.5);
  const elongation = elongationUnits / PHASE_QUANTIZATION_SCALE;
  const boundaryDistanceUnits = Math.min(
    elongationUnits,
    Math.abs(elongationUnits - HALF_CIRCLE_UNITS),
    Math.abs(FULL_CIRCLE_UNITS - elongationUnits),
  );
  if (boundaryDistanceUnits <= PHASE_GUARD_UNITS) return ['unknown', elongation];
  return [elongationUnits < HALF_CIRCLE_UNITS ? 'benefic' : 'malefic', elongation];
}

export function classifyNaturalGrahaNatures(
  chart: ElectionChartSnapshot,
): NaturalGrahaNatureResult {
  const positions = completePlanetPositions(chart, CANONICAL_PLANETS);
  if (!positions) {
    return {
      complete: false,
      natures: unknownNatures(),
      evidence: ['Complete canonical nine-graha facts are unavailable or invalid.'],
    };
  }

  const natures = unknownNatures();
  for (const name of FIXED_MALEFICS) natures[name] = 'malefic';
  for (const name of FIXED_BENEFICS) natures[name] = 'benefic';
  const [moonNature, elongation] = resolvedChandraNature(positions);
  natures.Chandra = moonNature;
  const guard = NATURAL_GRAHA_NATURE_CONVENTION.chandra_phase_guard_degrees;
  const chandraEvidence = moonNature === 'unknown'
    ? `Chandra elongation ${elongation.toFixed(4)}° is inside the inclusive ${guard.toFixed(2)}° phase-boundary guard.`
    : `Chandra elongation ${elongation.toFixed(4)}° resolves ${moonNature}.`;

  const budha = positions.get('Budha')!;
  const companions = CANONICAL_PLANET_ORDER.filter(name =>
    name !== 'Budha' && positions.get(name)?.rashi === budha.rashi);
  const maleficWitnesses = companions.filter(name => natures[name] === 'malefic');
  let budhaEvidence: string;
  if (maleficWitnesses.length) {
    natures.Budha = 'malefic';
    budhaEvidence = `Budha shares ${budha.rashi} with resolved malefic ${maleficWitnesses.join(', ')}.`;
  } else if (companions.includes('Chandra') && natures.Chandra === 'unknown') {
    natures.Budha = 'unknown';
    budhaEvidence = `Budha shares ${budha.rashi} only with phase-unknown Chandra as a possible malefic witness.`;
  } else {
    natures.Budha = 'benefic';
    budhaEvidence = `No resolved malefic shares Budha's sidereal Rasi ${budha.rashi}.`;
  }
  return {
    complete: true,
    natures: Object.fromEntries(CANONICAL_PLANET_ORDER.map(name => [name, natures[name]])),
    evidence: [chandraEvidence, budhaEvidence],
  };
}

function normalizedHouseSet(houses: unknown): ReadonlySet<number> | null {
  if (!Array.isArray(houses) || !houses.length) return null;
  if (houses.some(house =>
    !Number.isInteger(house) || house < 1 || house > 12)) return null;
  return new Set(houses);
}

function houseWitnesses(
  chart: ElectionChartSnapshot,
  houses: unknown,
): { result: NaturalGrahaNatureResult; occupants: Array<[string, number]> } | null {
  const targetHouses = normalizedHouseSet(houses);
  if (!targetHouses) return null;
  const result = classifyNaturalGrahaNatures(chart);
  if (!result.complete) return { result, occupants: [] };
  const positions = completePlanetPositions(chart, CANONICAL_PLANETS);
  if (!positions) return { result: { ...result, complete: false }, occupants: [] };
  return {
    result,
    occupants: CANONICAL_PLANET_ORDER
      .map(name => [name, positions.get(name)?.house] as const)
      .filter((item): item is [string, number] =>
        item[1] !== undefined && targetHouses.has(item[1])),
  };
}

export function evaluateExistentialBeneficHouseSet(
  chart: ElectionChartSnapshot,
  houses: readonly number[],
): PrimitiveOutcome {
  const witnesses = houseWitnesses(chart, houses);
  if (!witnesses) {
    return { status: 'unknown', evidence: ['The required house set is invalid.'] };
  }
  const { result, occupants } = witnesses;
  if (!result.complete) return { status: 'unknown', evidence: result.evidence };
  const benefics = occupants.filter(([name]) => result.natures[name] === 'benefic');
  if (benefics.length) {
    const detail = benefics.map(([name, house]) => `${name} (house ${house})`).join(', ');
    return { status: 'pass', evidence: [`Resolved natural-benefic witness: ${detail}.`] };
  }
  const unresolved = occupants.filter(([name]) => result.natures[name] === 'unknown');
  if (unresolved.length) {
    const detail = unresolved.map(([name, house]) => `${name} (house ${house})`).join(', ');
    return { status: 'unknown', evidence: [`Natural-benefic witness remains unresolved: ${detail}.`] };
  }
  return { status: 'fail', evidence: ['No resolved natural benefic occupies the required house set.'] };
}

export function evaluateForbiddenMaleficHouseSet(
  chart: ElectionChartSnapshot,
  houses: readonly number[],
): PrimitiveOutcome {
  const witnesses = houseWitnesses(chart, houses);
  if (!witnesses) {
    return { status: 'unknown', evidence: ['The forbidden house set is invalid.'] };
  }
  const { result, occupants } = witnesses;
  if (!result.complete) return { status: 'unknown', evidence: result.evidence };
  const malefics = occupants.filter(([name]) => result.natures[name] === 'malefic');
  if (malefics.length) {
    const detail = malefics.map(([name, house]) => `${name} (house ${house})`).join(', ');
    return { status: 'fail', evidence: [`Forbidden natural-malefic witness: ${detail}.`] };
  }
  const unresolved = occupants.filter(([name]) => result.natures[name] === 'unknown');
  if (unresolved.length) {
    const detail = unresolved.map(([name, house]) => `${name} (house ${house})`).join(', ');
    return { status: 'unknown', evidence: [`Forbidden-malefic check remains unresolved: ${detail}.`] };
  }
  return { status: 'pass', evidence: ['No resolved natural malefic occupies the forbidden house set.'] };
}

export function evaluateHouseFreeOfNaturalMalefics(
  rule: ElectionPrimitiveRule,
  positions: ReadonlyMap<string, PlanetPosition> | null,
  options: { houseFrameUncertain?: boolean } = {},
): PrimitiveOutcome {
  if (!positions) {
    return { status: 'unknown', evidence: [COMPLETE_GRAHA_FACTS_UNAVAILABLE] };
  }
  if (options.houseFrameUncertain) {
    return { status: 'unknown', evidence: ['The local-Lagna house frame is uncertain.'] };
  }

  const fixed = (rule.fixed_malefics || []).filter(
    name => positions.get(name)?.house === rule.house,
  );
  if (fixed.length) {
    return {
      status: 'fail',
      evidence: [`Natural malefics in Lagna: ${fixed.join(', ')}.`],
    };
  }

  const chandra = positions.get('Chandra');
  const surya = positions.get('Surya');
  if (!chandra || !surya) {
    return { status: 'unknown', evidence: [COMPLETE_GRAHA_FACTS_UNAVAILABLE] };
  }
  if (chandra.house !== rule.house) {
    return {
      status: 'pass',
      evidence: ['Natural malefics in Lagna: none; Chandra is outside Lagna.'],
    };
  }

  const elongation = ((longitude(chandra) - longitude(surya)) % 360 + 360) % 360;
  const guard = rule.lunar_phase_guard_degrees
    ?? NATURAL_MALEFIC_LUNAR_PHASE_GUARD_DEGREES;
  const boundaryDistance = Math.min(
    elongation, 360 - elongation, Math.abs(elongation - 180),
  );
  if (boundaryDistance <= guard + BOUNDARY_EPSILON) {
    return {
      status: 'unknown',
      evidence: [
        `Chandra occupies Lagna at ${elongation.toFixed(2)}° solar elongation, inside the disclosed ±${guard.toFixed(2)}° phase boundary guard.`,
      ],
    };
  }
  if (elongation > 180) {
    return {
      status: 'fail',
      evidence: [
        `Natural malefics in Lagna: waning Chandra (${elongation.toFixed(2)}° solar elongation).`,
      ],
    };
  }
  return {
    status: 'pass',
    evidence: [
      `Natural malefics in Lagna: none; waxing Chandra (${elongation.toFixed(2)}° solar elongation) is not malefic under this convention.`,
    ],
  };
}
